mono/metadata/object.c

```diff
@@ -101,6 +101,56 @@ mono_runtime_object_init (MonoObject *this_obj)
 	mono_error_assert_ok (error);
 }
 
+/**
+ * mono_runtime_object_init_checked:
+ * \param this_obj the object to initialize
+ * \param error set on error.
+ * This function calls the zero-argument constructor (which must
+ * exist) for the given object and returns TRUE on success, or FALSE
+ * on error and sets \p error.
+ */
+gboolean
+mono_runtime_object_init_handle (MonoObjectHandle this_obj, MonoError *error)
+{
+	MONO_REQ_GC_UNSAFE_MODE;
+
+	HANDLE_FUNCTION_ENTER ()
+
+	MonoMethod *method = NULL;
+	MonoClass *klass = MONO_HANDLE_GETVAL (this_obj, vtable)->klass;
+
+	error_init (error);
+	method = mono_class_get_method_from_name (klass, ".ctor", 0);
+	g_assertf (method, "Could not lookup zero argument constructor for class %s", mono_type_get_full_name (klass));
+
+	guint gchandle = 0;
+	gpointer raw;
+
+	if (method->klass->valuetype)
+		raw = mono_object_handle_pin_unbox (this_obj, &gchandle);
+	else
+		raw = MONO_HANDLE_RAW (this_obj);
+
+	mono_runtime_invoke_checked (method, raw, NULL, error);
```

Comment:
This is a pretty sketchy suggestion, but maybe it's alright not to pin the valuetype `this` during the invoke... once we're in managed it'll be kept alive because it's a root. and uh... _handwavy argument that it'll be alright until the native-to-managed transition_.

I'm really uncertain about this.

Reply:
Since it sketchy, ignore it, keep my code, which I believe is faithful to the prior code? Revisit another time, every invoke helper?

Reply:
Oh, hm, I was replacing mono_object_unbox with mono_object_handle_pin_unbox. I need to see if that the closest or exact analog.

Reply:
This does appear heavy handed in terms of what I was trying to do.

Reply:
Oh, the original did not pin. So I removed the pinning here too.